sw/device/tests/entropy_src_fw_observe_many_contiguous.c

```diff
@@ -0,0 +1,170 @@
+          .buffer_threshold = kEntropySrcFifoThreshold,
+      },
+      kDifToggleEnabled));
+  CHECK_DIF_OK(dif_entropy_src_configure(
+      entropy_src,
+      (dif_entropy_src_config_t){
+          .fips_enable = true,
+          .route_to_firmware = false,
+          .bypass_conditioner = false,
+          .single_bit_mode = kDifEntropySrcSingleBitModeDisabled,
+          .health_test_threshold_scope = false,
+          .health_test_window_size = kEntropySrcHealthTestWindowSize,
+          .alert_threshold = UINT16_MAX},
+      kDifToggleEnabled));
+
+  while (repeat_count-- > 0) {
+    LOG_INFO("collecting %u samples...", nr_samples);
+    // Collect samples from the the observe FIFO.
+    uint32_t words_to_read = ceil_div(nr_samples * 4, sizeof(uint32_t));
+    uint32_t *sample_buffer_ptr = sample_buffer;
+    ibex_timeout_t tmo = ibex_timeout_init(timeout_usec);
+    while(words_to_read > 0 && !ibex_timeout_check(&tmo)) {
+      uint32_t rem = words_to_read;
+      if (rem > kEntropySrcFifoThreshold) {
+        rem = kEntropySrcFifoThreshold;
+      }
+      CHECK_DIF_OK(dif_entropy_src_observe_fifo_blocking_read(
+          entropy_src, sample_buffer_ptr, rem));
+      sample_buffer_ptr += rem;
+      words_to_read -= rem;
+    }
+    CHECK(!ibex_timeout_check(&tmo), "did not collect samples in time");
+    // Make sure the FIFO did not overflow.
+    bool has_overflowed;
+    CHECK_DIF_OK(dif_entropy_src_has_fifo_overflowed(entropy_src, &has_overflowed));
+    CHECK(!has_overflowed, "observe FIFO overflowed during collection");
+    uint64_t elapsed = ibex_timeout_elapsed(&tmo);
+    uint64_t freq = udiv64_slow((uint64_t)nr_samples * (uint64_t)1000000, elapsed, NULL);
+    LOG_INFO("done in %ums (~ %usamples/s)", (uint32_t)udiv64_slow(elapsed, 1000, NULL), (uint32_t)freq);
+
+    // Let observe FIFO overflow
+    if (repeat_count > 0) {
+      LOG_INFO("let observe FIFO overflow...");
+      ibex_timeout_t tmo = ibex_timeout_init(timeout_usec);
+      bool has_overflowed = false;
+      while(!has_overflowed && !ibex_timeout_check(&tmo)) {
+        CHECK_DIF_OK(dif_entropy_src_has_fifo_overflowed(entropy_src, &has_overflowed));
+      }
+      CHECK(!ibex_timeout_check(&tmo), "observe FIFO did not overflow in time");
+      uint64_t elapsed = ibex_timeout_elapsed(&tmo);
+      LOG_INFO("done in %ums", (uint32_t)udiv64_slow(elapsed, 1000, NULL));
+      // Drain observe FIFO.
+      LOG_INFO("drain observe FIFO overflow...");
+      tmo = ibex_timeout_init(timeout_usec);
+      uint32_t words_to_read;
+      CHECK_DIF_OK(dif_entropy_src_get_fifo_depth(entropy_src, &words_to_read));
+      while(words_to_read > 0 && !ibex_timeout_check(&tmo)) {
+        uint32_t rem = words_to_read;
+        if (rem > kEntropySrcFifoThreshold) {
+          rem = kEntropySrcFifoThreshold;
+        }
+        CHECK_DIF_OK(dif_entropy_src_observe_fifo_blocking_read(
+            entropy_src, NULL, rem));
+        words_to_read -= rem;
+      }
```

Comment:
I am not sure it this is going to work and if yes, it's really robust. Because `dif_entropy_src_observe_fifo_blocking_read()` embeds a check that `rem` is less than or equal to the FIFO threshold size. Right now threshold size is half the FIFO depth.

Do we maybe need a function in the DIF to just read and empty the FIFO without caring about interrupts / thresholds etc.?

Reply:
Should instead `dif_entropy_src_observe_fifo_blocking_read` be changed to not check against the threshold? I don't think there is a good reason for this check.